sling/nlp/parser/trainer/transition-generator.cc

```diff
@@ -0,0 +1,457 @@
+  FrameInfo::Entry *frame = nullptr;     // frame responsible for the action
+  FrameInfo::Entry *neighbor = nullptr;  // neighboring frame (if applicable)
+};  
+
+// Map from token -> mentions starting/ending/both at it.
+struct TokenToMentions {
+  void Start(const Span *span) {
+    if (!starting.empty()) {
+      QCHECK_GE(starting.back()->end(), span->end());
+    } 
+    starting.emplace_back(span);
+  }
+    
+  void End(const Span *span) {
+    if (!ending.empty()) { 
+      QCHECK_LE(ending.front()->begin(), span->begin());
+    }   
+    ending.insert(ending.begin(), span);
+  }
+
+  void Singleton(const Span *span) {
+    singletons.emplace_back(span);
+  }
+
+  bool empty() const {
+    return starting.empty() && ending.empty() && singletons.empty();
+  }
+
+  std::vector<const Span *> starting;
+  std::vector<const Span *> ending;
+  std::vector<const Span *> singletons;
+};
+
+int Index(const Handles &attention, Handle item) {
+  for (int i = 0; i < attention.size(); ++i) {
+    if (attention[i] == item) return i;
+  }
+  return -1;
+}
+
+// Translates 'action' into an attention-resolved version.
+ParserAction Translate(const Handles &attention, const Action &action) {
+  ParserAction output(action.core.type);
+  if (action.core.length != 0) output.length = action.core.length;
+  if (!action.core.role.IsNil()) output.role = action.core.role;
+
+  switch (output.type) {
+    case ParserAction::EVOKE:
+      output.label = action.frame->type;
+      break;
+    case ParserAction::REFER:
+      output.target = Index(attention, action.frame->handle);
+      break;
+    case ParserAction::EMBED:
+      output.label = action.frame->type;
+      output.target = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::ELABORATE:
+      output.label = action.frame->type;
+      output.source = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::CONNECT:
+      output.source = Index(attention, action.frame->handle);
+      output.target = Index(attention, action.neighbor->handle);
+      break;
+    case ParserAction::ASSIGN:
+      output.source = Index(attention, action.frame->handle);
+      output.label = action.core.label;
+      break;
+    default:
+    break;
+  }
+
+  return output;
+}
+
+// Updates 'attention' as a result of execution 'action'.
+void Update(const Action &action, Handles *attention) {
+  auto type = action.core.type;
+  if (type == ParserAction::EVOKE || type == ParserAction::EMBED ||
+      type == ParserAction::ELABORATE) {
+    attention->insert(attention->begin(), action.frame->handle);
+  } else if (type == ParserAction::REFER || type == ParserAction::ASSIGN ||
+             type == ParserAction::CONNECT) {
+    std::remove(attention->begin(), attention->end(), action.frame->handle);
+    attention->insert(attention->begin(), action.frame->handle);
+  }
+}
+
+void InitInfo(
+    Frame frame, FrameInfo *frame_info, HandleSet *initialized) {
```

Comment:
Do you need to have frame by value here? Could you use "const Frame &frame" instead?

Reply:
Done